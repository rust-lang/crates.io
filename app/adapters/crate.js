import ApplicationAdapter from './application';

const BULK_REQUEST_GROUP_SIZE = 10;

export default class CrateAdapter extends ApplicationAdapter {
  coalesceFindRequests = true;

  findRecord(store, type, id, snapshot) {
    return super.findRecord(store, type, id, setDefaultInclude(snapshot));
  }

  queryRecord(store, type, query, adapterOptions) {
    return super.queryRecord(store, type, setDefaultInclude(query), adapterOptions);
  }

  /** Removes the `name` query parameter and turns it into a path parameter instead */
  urlForQueryRecord(query) {
    let baseUrl = super.urlForQueryRecord(...arguments);
    if (!query.name) {
      return baseUrl;
    }

    let crateName = query.name;
    delete query.name;
    return `${baseUrl}/${crateName}`;
  }

  /** Adds a `message` query parameter to the URL, if set in the `adapterOptions`. */
  urlForDeleteRecord(id, modelName, snapshot) {
    let url = super.urlForDeleteRecord(...arguments);

    let message = snapshot.adapterOptions.message;
    if (message) {
      url += `?message=${encodeURIComponent(message)}`;
    }

    return url;
  }

  groupRecordsForFindMany(store, snapshots) {
    let result = [];
    for (let i = 0; i < snapshots.length; i += BULK_REQUEST_GROUP_SIZE) {
      result.push(snapshots.slice(i, i + BULK_REQUEST_GROUP_SIZE));
    }
    return result;
  }
}

function setDefaultInclude(query) {
  if (query.include === undefined) {
    // This ensures `crate.versions` are always fetched from another request.
    query.include = 'keywords,categories,downloads';
  }

  return query;
}
