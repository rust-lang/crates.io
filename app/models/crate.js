import Model, { attr, hasMany } from '@ember-data/model';
import { assert } from '@ember/debug';
import { waitForPromise } from '@ember/test-waiters';
import { cached } from '@glimmer/tracking';

import { apiAction } from '@mainmatter/ember-api-actions';
import { task } from 'ember-concurrency';

export default class Crate extends Model {
  @attr name;
  @attr downloads;
  @attr recent_downloads;
  @attr('date') created_at;
  @attr('date') updated_at;
  /**
   * This is the default version that will be shown when visiting the crate
   * details page. Note that this value can be `null`, which may be unexpected.
   * @type {string | null}
   */
  @attr default_version;
  @attr yanked;
  @attr max_version;
  @attr max_stable_version;
  @attr newest_version;

  @attr description;
  @attr homepage;
  @attr documentation;
  @attr repository;

  @hasMany('version', { async: true, inverse: 'crate' }) versions;
  @hasMany('team', { async: true, inverse: null }) owner_team;
  @hasMany('user', { async: true, inverse: null }) owner_user;
  @hasMany('version-download', { async: true, inverse: null }) version_downloads;
  @hasMany('keyword', { async: true, inverse: null }) keywords;
  @hasMany('category', { async: true, inverse: null }) categories;
  @hasMany('dependency', { async: true, inverse: null }) reverse_dependencies;

  @cached get versionIdsBySemver() {
    let { last } = this.loadVersionsTask;
    assert('`loadVersionsTask.perform()` must be called before calling `versionIdsBySemver`', last != null);
    let versions = last?.value ?? [];
    return versions
      .slice()
      .sort(compareVersionBySemver)
      .map(v => v.id);
  }

  @cached get versionIdsByDate() {
    let { last } = this.loadVersionsTask;
    assert('`loadVersionsTask.perform()` must be called before calling `versionIdsByDate`', last != null);
    let versions = last?.value ?? [];
    return versions
      .slice()
      .sort(compareVersionByDate)
      .map(v => v.id);
  }

  @cached get firstVersionId() {
    return this.versionIdsByDate.at(-1);
  }

  @cached get versionsObj() {
    let { last } = this.loadVersionsTask;
    assert('`loadVersionsTask.perform()` must be called before calling `versionsObj`', last != null);
    let versions = last?.value ?? [];
    return Object.fromEntries(versions.slice().map(v => [v.id, v]));
  }

  @cached get releaseTrackSet() {
    let map = new Map();
    let { versionsObj: versions, versionIdsBySemver } = this;
    for (let id of versionIdsBySemver) {
      let { releaseTrack, isPrerelease, yanked } = versions[id];
      if (releaseTrack && !isPrerelease && !yanked && !map.has(releaseTrack)) {
        map.set(releaseTrack, id);
      }
    }
    return new Set(map.values());
  }

  hasOwnerUser(userId) {
    let { last } = this.loadOwnerUserTask;
    assert('`loadOwnerUserTask.perform()` must be called before calling `hasOwnerUser()`', last != null);
    return (last?.value ?? []).some(({ id }) => id === userId);
  }

  get owners() {
    let { last } = this.loadOwnersTask;
    assert('`loadOwnersTask.perform()` must be called before accessing `owners`', last != null);
    return last?.value ?? [];
  }

  async follow() {
    return await waitForPromise(apiAction(this, { method: 'PUT', path: 'follow' }));
  }

  async unfollow() {
    return await waitForPromise(apiAction(this, { method: 'DELETE', path: 'follow' }));
  }

  async inviteOwner(username) {
    let response = await waitForPromise(
      apiAction(this, { method: 'PUT', path: 'owners', data: { owners: [username] } }),
    );
    if (response.ok) {
      return response;
    } else {
      throw response;
    }
  }

  async removeOwner(username) {
    let response = await waitForPromise(
      apiAction(this, { method: 'DELETE', path: 'owners', data: { owners: [username] } }),
    );
    if (response.ok) {
      return response;
    } else {
      throw response;
    }
  }

  loadOwnerUserTask = task(async () => {
    return (await this.owner_user) ?? [];
  });

  loadOwnersTask = task(async () => {
    let [teams, users] = await Promise.all([this.owner_team, this.owner_user]);
    return [...(teams ?? []), ...(users ?? [])];
  });

  loadVersionsTask = task(async ({ reload = false } = {}) => {
    let versionsRef = this.hasMany('versions');
    let fut = reload === true ? versionsRef.reload() : versionsRef.load();
    return (await fut) ?? [];
  });
}

function compareVersionBySemver(a, b) {
  let aSemver = a.semver;
  let bSemver = b.semver;

  if (aSemver === bSemver) {
    return b.created_at - a.created_at;
  } else if (aSemver === null) {
    return 1;
  } else if (bSemver === null) {
    return -1;
  } else {
    return bSemver.compare(aSemver);
  }
}

function compareVersionByDate(a, b) {
  let bDate = b.created_at.getTime();
  let aDate = a.created_at.getTime();

  return bDate === aDate ? parseInt(b.id) - parseInt(a.id) : bDate - aDate;
}
