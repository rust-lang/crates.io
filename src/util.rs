pub use self::io_util::{read_fill, read_le_u32};
pub use self::request_helpers::*;
pub use crates_io_database::utils::token;

pub mod diesel;
pub mod errors;
mod io_util;
mod request_helpers;
pub mod string_excl_null;
pub mod tracing;
