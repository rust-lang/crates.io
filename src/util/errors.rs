//! This module implements several error types and traits.
//!
//! The suggested usage in returned results is as follows:
//!
//! * The concrete `util::concrete::Error` type (re-exported as `util::Error`) is great for code
//!   that is not part of the request/response lifecycle.  It avoids pulling in the unnecessary
//!   infrastructure to convert errors into a user facing JSON responses (relative to `AppError`).
//! * `diesel::QueryResult` - There is a lot of code that only deals with query errors.  If only
//!   one type of error is possible in a function, using that specific error is preferable to the
//!   more general `util::Error`.  This is especially common in model code.
//! * `util::errors::AppResult` - Some failures should be converted into user facing JSON
//!   responses.  This error type is more dynamic and is box allocated.  Low-level errors are
//!   typically not converted to user facing errors and most usage is within the models,
//!   controllers, and middleware layers.

use axum::response::IntoResponse;
use std::any::{Any, TypeId};
use std::borrow::Cow;
use std::error::Error;
use std::fmt;

use axum::Extension;
use chrono::NaiveDateTime;
use diesel::result::{DatabaseErrorKind, Error as DieselError};
use http::StatusCode;
use tokio::task::JoinError;

use crate::middleware::log_request::ErrorField;

mod json;

use crate::email::EmailError;
use crate::util::diesel::is_read_only_error;
use crates_io_github::GitHubError;
pub use json::TOKEN_FORMAT_ERROR;
pub(crate) use json::{custom, InsecurelyGeneratedTokenRevoked, TooManyRequests};

pub type BoxedAppError = Box<dyn AppError>;

// The following are intended to be used for errors being sent back to the Ember
// frontend, not to cargo as cargo does not handle non-200 response codes well
// (see <https://github.com/rust-lang/cargo/issues/3995>), but Ember requires
// non-200 response codes for its stores to work properly.

/// Return an error with status 400 and the provided description as JSON
pub fn bad_request<S: ToString>(error: S) -> BoxedAppError {
    custom(StatusCode::BAD_REQUEST, error.to_string())
}

pub fn account_locked(reason: &str, until: Option<NaiveDateTime>) -> BoxedAppError {
    let detail = until
        .map(|until| until.format("%F at %T UTC"))
        .map(|until| format!("This account is locked until {until}. Reason: {reason}"))
        .unwrap_or_else(|| format!("This account is indefinitely locked. Reason: {reason}"));

    custom(StatusCode::FORBIDDEN, detail)
}

pub fn forbidden(detail: impl Into<Cow<'static, str>>) -> BoxedAppError {
    custom(StatusCode::FORBIDDEN, detail)
}

pub fn not_found() -> BoxedAppError {
    custom(StatusCode::NOT_FOUND, "Not Found")
}

/// Returns an error with status 500 and the provided description as JSON
pub fn server_error<S: ToString>(error: S) -> BoxedAppError {
    custom(StatusCode::INTERNAL_SERVER_ERROR, error.to_string())
}

/// Returns an error with status 503 and the provided description as JSON
pub fn service_unavailable() -> BoxedAppError {
    custom(StatusCode::SERVICE_UNAVAILABLE, "Service unavailable")
}

pub fn crate_not_found(krate: &str) -> BoxedAppError {
    let detail = format!("crate `{krate}` does not exist");
    custom(StatusCode::NOT_FOUND, detail)
}

pub fn version_not_found(krate: &str, version: &str) -> BoxedAppError {
    let detail = format!("crate `{krate}` does not have a version `{version}`");
    custom(StatusCode::NOT_FOUND, detail)
}

// =============================================================================
// AppError trait

pub trait AppError: Send + fmt::Display + fmt::Debug + 'static {
    /// Generate an HTTP response for the error
    ///
    /// If none is returned, the error will bubble up the middleware stack
    /// where it is eventually logged and turned into a status 500 response.
    fn response(&self) -> axum::response::Response;

    fn get_type_id(&self) -> TypeId {
        TypeId::of::<Self>()
    }
}

impl dyn AppError {
    pub fn is<T: Any>(&self) -> bool {
        self.get_type_id() == TypeId::of::<T>()
    }
}

impl AppError for BoxedAppError {
    fn response(&self) -> axum::response::Response {
        (**self).response()
    }

    fn get_type_id(&self) -> TypeId {
        (**self).get_type_id()
    }
}

impl IntoResponse for BoxedAppError {
    fn into_response(self) -> axum::response::Response {
        self.response()
    }
}

pub type AppResult<T> = Result<T, BoxedAppError>;

// =============================================================================
// Error impls

impl<E: Error + Send + 'static> AppError for E {
    fn response(&self) -> axum::response::Response {
        error!(error = %self, "Internal Server Error");

        sentry::capture_error(self);

        server_error_response(self.to_string())
    }
}

impl From<diesel::ConnectionError> for BoxedAppError {
    fn from(err: diesel::ConnectionError) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<DieselError> for BoxedAppError {
    fn from(err: DieselError) -> BoxedAppError {
        match err {
            DieselError::NotFound => not_found(),
            e if is_read_only_error(&e) => {
                let detail = "crates.io is currently in read-only mode. Please check https://status.crates.io/ for details and try again later.";
                custom(StatusCode::SERVICE_UNAVAILABLE, detail)
            }
            DieselError::DatabaseError(DatabaseErrorKind::ClosedConnection, _) => {
                service_unavailable()
            }
            _ => Box::new(err),
        }
    }
}

impl From<EmailError> for BoxedAppError {
    fn from(error: EmailError) -> Self {
        match error {
            EmailError::AddressError(error) => Box::new(error),
            EmailError::MessageBuilderError(error) => Box::new(error),
            EmailError::TransportError(error) => {
                error!(?error, "Failed to send email");
                server_error("Failed to send the email")
            }
        }
    }
}

impl From<diesel_async::pooled_connection::deadpool::PoolError> for BoxedAppError {
    fn from(err: diesel_async::pooled_connection::deadpool::PoolError) -> BoxedAppError {
        error!("Database pool error: {err}");
        service_unavailable()
    }
}

impl From<prometheus::Error> for BoxedAppError {
    fn from(err: prometheus::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<reqwest::Error> for BoxedAppError {
    fn from(err: reqwest::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<serde_json::Error> for BoxedAppError {
    fn from(err: serde_json::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<std::io::Error> for BoxedAppError {
    fn from(err: std::io::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<crates_io_worker::EnqueueError> for BoxedAppError {
    fn from(err: crates_io_worker::EnqueueError) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<JoinError> for BoxedAppError {
    fn from(err: JoinError) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<GitHubError> for BoxedAppError {
    fn from(error: GitHubError) -> Self {
        match error {
            GitHubError::Permission(_) => custom(
                StatusCode::FORBIDDEN,
                "It looks like you don't have permission \
                     to query a necessary property from GitHub \
                     to complete this request. \
                     You may need to re-authenticate on \
                     crates.io to grant permission to read \
                     GitHub org memberships.",
            ),
            GitHubError::NotFound(_) => not_found(),
            _ => internal(format!("didn't get a 200 result from github: {error}")),
        }
    }
}

// =============================================================================
// Internal error for use with `chain_error`

#[derive(Debug)]
struct InternalAppError {
    description: String,
}

impl fmt::Display for InternalAppError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.description)?;
        Ok(())
    }
}

impl AppError for InternalAppError {
    fn response(&self) -> axum::response::Response {
        error!(error = %self.description, "Internal Server Error");

        sentry::capture_message(&self.description, sentry::Level::Error);

        server_error_response(self.description.to_string())
    }
}

pub fn internal<S: ToString>(error: S) -> BoxedAppError {
    Box::new(InternalAppError {
        description: error.to_string(),
    })
}

fn server_error_response(error: String) -> axum::response::Response {
    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Extension(ErrorField(error)),
        "Internal Server Error",
    )
        .into_response()
}

#[cfg(test)]
mod tests {
    use super::*;
    use diesel::result::Error as DieselError;
    use http::StatusCode;

    #[test]
    fn http_error_responses() {
        use crate::serde::de::Error;

        // Types for handling common error status codes
        assert_eq!(bad_request("").response().status(), StatusCode::BAD_REQUEST);
        assert_eq!(forbidden("").response().status(), StatusCode::FORBIDDEN);
        assert_eq!(
            BoxedAppError::from(DieselError::NotFound)
                .response()
                .status(),
            StatusCode::NOT_FOUND
        );
        assert_eq!(not_found().response().status(), StatusCode::NOT_FOUND);

        // All other error types are converted to internal server errors
        assert_eq!(
            internal("").response().status(),
            StatusCode::INTERNAL_SERVER_ERROR
        );
        assert_eq!(
            BoxedAppError::from(serde_json::Error::custom("ExpectedColon"))
                .response()
                .status(),
            StatusCode::INTERNAL_SERVER_ERROR
        );
        assert_eq!(
            BoxedAppError::from(::std::io::Error::new(::std::io::ErrorKind::Other, ""))
                .response()
                .status(),
            StatusCode::INTERNAL_SERVER_ERROR
        );
    }
}
