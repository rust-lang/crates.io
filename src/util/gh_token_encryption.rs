use aes_gcm::aead::{Aead, AeadCore, OsRng};
use aes_gcm::{Aes256Gcm, Key, KeyInit, Nonce};
use anyhow::{Context, Result};
use oauth2::AccessToken;

/// A struct that encapsulates GitHub token encryption and decryption
/// using AES-256-GCM.
pub struct GitHubTokenEncryption {
    cipher: Aes256Gcm,
}

impl GitHubTokenEncryption {
    /// Creates a new [GitHubTokenEncryption] instance with the provided cipher
    pub fn new(cipher: Aes256Gcm) -> Self {
        Self { cipher }
    }

    /// Creates a new [GitHubTokenEncryption] instance with a cipher for testing
    /// purposes.
    #[cfg(test)]
    pub fn for_testing() -> Self {
        let test_key = b"test_key_32_bytes_long_for_tests";
        Self::new(Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(test_key)))
    }

    /// Creates a new [GitHubTokenEncryption] instance from the environment
    ///
    /// Reads the `GITHUB_TOKEN_ENCRYPTION_KEY` environment variable, which
    /// should be a 64-character hex string (32 bytes when decoded).
    pub fn from_environment() -> Result<Self> {
        let gh_token_key = std::env::var("GITHUB_TOKEN_ENCRYPTION_KEY")
            .context("GITHUB_TOKEN_ENCRYPTION_KEY environment variable not set")?;

        if gh_token_key.len() != 64 {
            anyhow::bail!("GITHUB_TOKEN_ENCRYPTION_KEY must be exactly 64 hex characters");
        }

        let gh_token_key = hex::decode(gh_token_key.as_bytes())
            .context("GITHUB_TOKEN_ENCRYPTION_KEY must be exactly 64 hex characters")?;

        let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(&gh_token_key));

        Ok(Self::new(cipher))
    }

    /// Encrypts a GitHub access token using AES-256-GCM
    ///
    /// The encrypted data format is: `[12-byte nonce][encrypted data]`
    /// The nonce is randomly generated for each encryption to ensure uniqueness.
    pub fn encrypt(&self, plaintext: &str) -> Result<Vec<u8>> {
        // Generate a random nonce for this encryption
        let nonce = Aes256Gcm::generate_nonce(&mut OsRng);

        // Encrypt the token
        let encrypted = self
            .cipher
            .encrypt(&nonce, plaintext.as_bytes())
            .map_err(|error| anyhow::anyhow!("Failed to encrypt token: {error}"))?;

        // Combine nonce + ciphertext (which includes the auth tag)
        let mut result = Vec::with_capacity(nonce.len() + encrypted.len());
        result.extend_from_slice(&nonce);
        result.extend_from_slice(&encrypted);

        Ok(result)
    }

    /// Decrypts a GitHub access token using AES-256-GCM
    ///
    /// Expects the data format: `[12-byte nonce][encrypted data]`
    pub fn decrypt(&self, encrypted: &[u8]) -> Result<AccessToken> {
        if encrypted.len() < 12 {
            anyhow::bail!("Invalid encrypted token: too short");
        }

        // Extract nonce and ciphertext
        let (nonce_bytes, ciphertext) = encrypted.split_at(12);
        let nonce = Nonce::from_slice(nonce_bytes);

        // Decrypt the token
        let plaintext = self
            .cipher
            .decrypt(nonce, ciphertext)
            .context("Failed to decrypt token")?;

        let plaintext =
            String::from_utf8(plaintext).context("Decrypted token is not valid UTF-8")?;

        Ok(AccessToken::new(plaintext))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use aes_gcm::{Key, KeyInit};
    use claims::{assert_err, assert_ok};
    use insta::assert_snapshot;

    fn create_test_encryption() -> GitHubTokenEncryption {
        let key = Key::<Aes256Gcm>::from_slice(b"test_master_key_32_bytes_long!!!");
        let cipher = Aes256Gcm::new(key);
        GitHubTokenEncryption { cipher }
    }

    #[test]
    fn test_encrypt_decrypt_roundtrip() {
        let encryption = create_test_encryption();
        let original_token = "ghs_test_token_123456789";

        // Encrypt the token
        let encrypted = assert_ok!(encryption.encrypt(original_token));

        // Decrypt it back
        let decrypted = assert_ok!(encryption.decrypt(&encrypted));

        assert_eq!(original_token, decrypted.secret());
    }

    #[test]
    fn test_encrypt_produces_different_ciphertext() {
        let encryption = create_test_encryption();
        let token = "ghs_test_token_123456789";

        // Encrypt the same token twice
        let encrypted1 = assert_ok!(encryption.encrypt(token));
        let encrypted2 = assert_ok!(encryption.encrypt(token));

        // Should produce different ciphertext due to random nonce
        assert_ne!(encrypted1, encrypted2);

        // But both should decrypt to the same plaintext
        let decrypted1 = assert_ok!(encryption.decrypt(&encrypted1));
        let decrypted2 = assert_ok!(encryption.decrypt(&encrypted2));

        assert_eq!(decrypted1.secret(), decrypted2.secret());
        assert_eq!(decrypted1.secret(), token);
    }

    #[test]
    fn test_invalid_encrypted_data() {
        let encryption = create_test_encryption();

        // Too short
        let err = assert_err!(encryption.decrypt(&[1, 2, 3]));
        assert_snapshot!(err, @"Invalid encrypted token: too short");

        // Invalid data
        let invalid_data = vec![0u8; 50];
        let err = assert_err!(encryption.decrypt(&invalid_data));
        assert_snapshot!(err, @"Failed to decrypt token");
    }

    #[test]
    fn test_different_keys() {
        let encryption1 = create_test_encryption();

        // Create a different encryption with a different key
        let key2 = Key::<Aes256Gcm>::from_slice(b"different_key_32_bytes_long!!!!!");
        let cipher2 = Aes256Gcm::new(key2);
        let encryption2 = GitHubTokenEncryption { cipher: cipher2 };

        let token = "ghs_test_token_123456789";

        // Encrypt with encryption1
        let encrypted = assert_ok!(encryption1.encrypt(token));

        // Try to decrypt with encryption2 (should fail)
        let err = assert_err!(encryption2.decrypt(&encrypted));
        assert_snapshot!(err, @"Failed to decrypt token");

        // But encryption1 should still work
        let decrypted = assert_ok!(encryption1.decrypt(&encrypted));
        assert_eq!(decrypted.secret(), token);
    }
}
