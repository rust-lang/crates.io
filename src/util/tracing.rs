use crates_io_env_vars::var;
use sentry::integrations::tracing::EventFilter;
use tracing::Level;
use tracing::Metadata;
use tracing_subscriber::filter::LevelFilter;
use tracing_subscriber::{prelude::*, EnvFilter, Layer};

/// Initializes the `tracing` logging framework.
///
/// Regular CLI output is influenced by the optional
/// [`RUST_LOG`](tracing_subscriber::filter::EnvFilter) environment variable
/// and is showing all `INFO` level events by default.
///
/// This function also sets up the Sentry error reporting integration for the
/// `tracing` framework, which is hardcoded to include all `INFO` level events.
pub fn init() {
    init_with_default_level(LevelFilter::INFO)
}

fn init_with_default_level(level: LevelFilter) {
    let env_filter = EnvFilter::builder()
        .with_default_directive(level.into())
        .from_env_lossy();

    let log_format = var("RUST_LOG_FORMAT")
        .inspect_err(|error| {
            warn!("Failed to read RUST_LOG_FORMAT, falling back to default: {error}")
        })
        .unwrap_or_default();

    let log_layer = match log_format.as_deref() {
        Some("json") => tracing_subscriber::fmt::layer()
            .json()
            .with_filter(env_filter)
            .boxed(),
        _ => tracing_subscriber::fmt::layer()
            .compact()
            .without_time()
            .with_filter(env_filter)
            .boxed(),
    };

    let sentry_layer = sentry::integrations::tracing::layer()
        .event_filter(event_filter)
        .with_filter(LevelFilter::INFO);

    tracing_subscriber::registry()
        .with(log_layer)
        .with(sentry_layer)
        .init();
}

pub fn event_filter(metadata: &Metadata<'_>) -> EventFilter {
    match metadata.level() {
        &Level::ERROR if metadata.target() == "http" => EventFilter::Breadcrumb,
        &Level::ERROR => EventFilter::Exception,
        &Level::WARN | &Level::INFO => EventFilter::Breadcrumb,
        &Level::DEBUG | &Level::TRACE => EventFilter::Ignore,
    }
}

/// Initializes the `tracing` logging framework for usage in tests.
pub fn init_for_test() {
    let env_filter = EnvFilter::builder()
        .with_default_directive(LevelFilter::DEBUG.into())
        .from_env_lossy()
        .add_directive("tokio_postgres=info".parse().unwrap());

    let _ = tracing_subscriber::fmt()
        .compact()
        .with_env_filter(env_filter)
        .with_test_writer()
        .try_init();
}
