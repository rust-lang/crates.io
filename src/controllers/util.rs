use crate::middleware::app::RequestApp;
use crate::middleware::log_request::RequestLogExt;
use crate::util::errors::{forbidden, AppResult};
use crate::util::BytesRequest;
use http::request::Parts;
use http::{header, Extensions, HeaderMap, HeaderValue, Method, Request, Uri, Version};

/// The Origin header (<https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin>)
/// is sent with CORS requests and POST requests, and indicates where the request comes from.
/// We don't want to accept authenticated requests that originated from other sites, so this
/// function returns an error if the Origin header doesn't match what we expect "this site" to
/// be: <https://crates.io> in production, or <http://localhost:port/> in development.
pub fn verify_origin(parts: &Parts) -> AppResult<()> {
    let headers = parts.headers();
    let allowed_origins = &parts.app().config.allowed_origins;

    let bad_origin = headers
        .get_all(header::ORIGIN)
        .iter()
        .find(|value| !allowed_origins.contains(value));

    if let Some(bad_origin) = bad_origin {
        let error_message =
            format!("only same-origin requests can be authenticated. got {bad_origin:?}");

        parts.request_log().add("cause", error_message);

        return Err(forbidden("invalid origin header"));
    }
    Ok(())
}

pub trait RequestPartsExt {
    fn method(&self) -> &Method;
    fn uri(&self) -> &Uri;
    fn version(&self) -> Version;
    fn headers(&self) -> &HeaderMap<HeaderValue>;
    fn extensions(&self) -> &Extensions;
}

impl RequestPartsExt for Parts {
    fn method(&self) -> &Method {
        &self.method
    }
    fn uri(&self) -> &Uri {
        &self.uri
    }
    fn version(&self) -> Version {
        self.version
    }
    fn headers(&self) -> &HeaderMap<HeaderValue> {
        &self.headers
    }
    fn extensions(&self) -> &Extensions {
        &self.extensions
    }
}

impl<B> RequestPartsExt for Request<B> {
    fn method(&self) -> &Method {
        self.method()
    }
    fn uri(&self) -> &Uri {
        self.uri()
    }
    fn version(&self) -> Version {
        self.version()
    }
    fn headers(&self) -> &HeaderMap<HeaderValue> {
        self.headers()
    }
    fn extensions(&self) -> &Extensions {
        self.extensions()
    }
}

impl RequestPartsExt for BytesRequest {
    fn method(&self) -> &Method {
        self.0.method()
    }
    fn uri(&self) -> &Uri {
        self.0.uri()
    }
    fn version(&self) -> Version {
        self.0.version()
    }
    fn headers(&self) -> &HeaderMap<HeaderValue> {
        self.0.headers()
    }
    fn extensions(&self) -> &Extensions {
        self.0.extensions()
    }
}
