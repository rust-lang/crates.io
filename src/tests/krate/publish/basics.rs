use crate::builders::{CrateBuilder, PublishBuilder};
use crate::util::{RequestHelper, TestApp};
use crates_io::schema::versions_published_by;
use diesel::{QueryDsl, RunQueryDsl};
use googletest::prelude::*;
use http::StatusCode;
use insta::{assert_json_snapshot, assert_snapshot};

#[tokio::test(flavor = "multi_thread")]
async fn new_krate() {
    let (app, _, user) = TestApp::full().with_user();

    let crate_to_publish = PublishBuilder::new("foo_new", "1.0.0");
    let response = user.publish_crate(crate_to_publish).await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".crate.created_at" => "[datetime]",
        ".crate.updated_at" => "[datetime]",
    });

    let crates = app.crates_from_index_head("foo_new");
    assert_json_snapshot!(crates);

    assert_snapshot!(app.stored_files().await.join("\n"), @r###"
    crates/foo_new/foo_new-1.0.0.crate
    index/fo/o_/foo_new
    rss/crates.xml
    rss/crates/foo_new.xml
    rss/updates.xml
    "###);

    app.db(|conn| {
        let email: String = versions_published_by::table
            .select(versions_published_by::email)
            .first(conn)
            .unwrap();
        assert_eq!(email, "foo@example.com");
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn new_krate_with_token() {
    let (app, _, _, token) = TestApp::full().with_token();

    let crate_to_publish = PublishBuilder::new("foo_new", "1.0.0");
    let response = token.publish_crate(crate_to_publish).await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".crate.created_at" => "[datetime]",
        ".crate.updated_at" => "[datetime]",
    });

    assert_snapshot!(app.stored_files().await.join("\n"), @r###"
    crates/foo_new/foo_new-1.0.0.crate
    index/fo/o_/foo_new
    rss/crates.xml
    rss/crates/foo_new.xml
    rss/updates.xml
    "###);
}

#[tokio::test(flavor = "multi_thread")]
async fn new_krate_weird_version() {
    let (app, _, _, token) = TestApp::full().with_token();

    let crate_to_publish = PublishBuilder::new("foo_weird", "0.0.0-pre");
    let response = token.publish_crate(crate_to_publish).await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".crate.created_at" => "[datetime]",
        ".crate.updated_at" => "[datetime]",
    });

    assert_snapshot!(app.stored_files().await.join("\n"), @r###"
    crates/foo_weird/foo_weird-0.0.0-pre.crate
    index/fo/o_/foo_weird
    rss/crates.xml
    rss/crates/foo_weird.xml
    rss/updates.xml
    "###);
}

#[tokio::test(flavor = "multi_thread")]
async fn new_krate_twice() {
    let (app, _, _, token) = TestApp::full().with_token();

    let crate_to_publish = PublishBuilder::new("foo_twice", "0.99.0");
    token.publish_crate(crate_to_publish).await.good();

    let crate_to_publish =
        PublishBuilder::new("foo_twice", "2.0.0").description("2.0.0 description");
    let response = token.publish_crate(crate_to_publish).await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".crate.created_at" => "[datetime]",
        ".crate.updated_at" => "[datetime]",
    });

    let crates = app.crates_from_index_head("foo_twice");
    assert_json_snapshot!(crates);

    assert_snapshot!(app.stored_files().await.join("\n"), @r###"
    crates/foo_twice/foo_twice-0.99.0.crate
    crates/foo_twice/foo_twice-2.0.0.crate
    index/fo/o_/foo_twice
    rss/crates.xml
    rss/crates/foo_twice.xml
    rss/updates.xml
    "###);
}

#[tokio::test(flavor = "multi_thread")]
async fn new_krate_duplicate_version() {
    let (app, _, user, token) = TestApp::full().with_token();

    app.db(|conn| {
        // Insert a crate directly into the database and then we'll try to publish the same version
        CrateBuilder::new("foo_dupe", user.as_model().id)
            .version("1.0.0")
            .expect_build(conn);
    });

    let crate_to_publish = PublishBuilder::new("foo_dupe", "1.0.0");
    let response = token.publish_crate(crate_to_publish).await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_json_snapshot!(response.json());

    assert_that!(app.stored_files().await, empty());
}
