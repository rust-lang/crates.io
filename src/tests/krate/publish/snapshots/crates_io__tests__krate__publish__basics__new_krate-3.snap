---
source: src/tests/krate/publish/basics.rs
expression: crates
---
[
  {
    "name": "foo_new",
    "vers": "1.0.0",
    "deps": [],
    "cksum": "696b8fc0b2d62c399459b2000dbcab7d9914e5dfafff974f53c1212e02f014bb",
    "features": {},
    "yanked": false
  }
]
